plugins/pkgtorrent/server/pkgtorrent-service.py

```diff
@@ -0,0 +1,349 @@
+                ('Pragma', 'no-cache'),
+                ('Expires', '0'),
+            ])
+            return []
+
+    # if we get this far, we should have a list of pathspecs to compare
+    # to the expected digest.
+    assert sha1('\n'.join(pathspecs)).hexdigest() == digest
+
+    # It's possible we get a storm of people suddenly asking for exactly the
+    # same set of files, so make them block and return the cached one.
+    open(digestfile, 'w').close()
+    files = []
+    pieces = []
+    lookup = []
+    n = 0
+    prefixes = set()
+    nullfile = None
+    nullusage = {}
+    for path, mtime in paths:
+        # these always start with / as they're relative to the root
+        path = path[1:]
+        cachepath = os.path.join(PIECECACHE, path, str(BLOCKSIZE))
+        splitpath = path.split(os.sep)
+        prefixes.add(splitpath[0])
+
+        valid_cache = False
+        if os.path.exists(cachepath):
+            try:
+                data = pickle.load(open(cachepath))
+                if data[0] == PROTOCOL:
+                    protocol, size, blocks, tail, cached_mtime = data
+                    valid_cache = protocol == PROTOCOL and cached_mtime == mtime
+                    padding = BLOCKSIZE - (size % BLOCKSIZE)
+                else:
+                    valid_cache = False
+            except NotImplementedError:
+                # TODO: figure out pickle and file exceptions
+                valid_cache = False
+        if not valid_cache:
+            fullpath = os.path.join(document_root, path)
+            size, blocks, tail = os.stat(fullpath).st_size, [], None
+            f = open(fullpath)
+            for i in xrange(int(size / BLOCKSIZE)):
+                m = sha1()
+                lookup.append((len(files), f.tell()))
+                m.update(f.read(BLOCKSIZE))
+                blocks.append(m.digest())
+            remainder = size % BLOCKSIZE
+            if remainder:
+                m = sha1()
+                # read as much as we can left
+                lookup.append((len(files), f.tell()))
+                buffer = f.read()
+                assert len(buffer) == remainder
+                m.update(buffer)
+                tail = m.digest()
+                padding = BLOCKSIZE - remainder
+                m.update('\0' * padding)
+                blocks.append(m.digest())
+            blocks = ''.join(blocks)
+            f.close()
+            cachedir = os.path.dirname(cachepath)
+            if not os.path.exists(cachedir):
+                os.makedirs(cachedir)
+            pickle.dump(
+                (PROTOCOL, size, blocks, tail, mtime),
+                open(cachepath, 'wb'),
+                pickle.HIGHEST_PROTOCOL,
+            )
+        # url-list requires a name, and that needs stripped
+        # off all the paths inside the torrent to work.
+        files.append({
+            'path': splitpath[1:],
+            'length': size,
+        })
+        if padding:
```

Comment:
`padding` may be undefined at this point; adding `padding = None` just before the for loop should solve that.